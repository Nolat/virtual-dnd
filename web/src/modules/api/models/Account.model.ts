import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  OneToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn
} from "typeorm";

import User from "./User.model";

@Entity()
export default class Account extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: string;

  @OneToOne(() => User)
  @JoinColumn()
  user: User;

  @Column()
  providerType: string;

  @Column()
  providerId: string;

  @Column()
  providerAccountId: string;

  @Column({ nullable: true })
  refreshToken: string;

  @Column({ nullable: true })
  accessToken: string;

  @Column({ type: "timestamp", nullable: true })
  accessTokenExpiresOn: Date;

  @CreateDateColumn({ type: "timestamp" })
  createdAt: Date;

  @UpdateDateColumn({ type: "timestamp" })
  updatedAt: Date;
}
