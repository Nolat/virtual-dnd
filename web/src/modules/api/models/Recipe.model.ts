import { Field, ID, ObjectType } from "type-graphql";
import { BaseEntity, Column, Entity, PrimaryGeneratedColumn } from "typeorm";

@ObjectType()
@Entity()
export class Recipe extends BaseEntity {
  @Field(() => ID)
  @PrimaryGeneratedColumn()
  id: string;

  @Field()
  @Column()
  title: string;

  @Field({ nullable: true })
  @Column({ nullable: true })
  description?: string;
}
