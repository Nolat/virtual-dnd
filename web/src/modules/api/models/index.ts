export { default as User } from "./User.model";
export { default as Account } from "./Account.model";
export { default as Session } from "./Session.model";
export { default as Game } from "./Game.model";
export { default as GameUser } from "./GameUser.model";
