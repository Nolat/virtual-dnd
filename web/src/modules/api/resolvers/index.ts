export { GameResolver } from "./Game.resolver";
export { UserResolver } from "./User.resolver";
