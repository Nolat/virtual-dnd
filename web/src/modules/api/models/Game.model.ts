import { Field, ID, ObjectType } from "type-graphql";
import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn
} from "typeorm";

import { GameUser, User } from "modules/api/models";

@Entity()
@ObjectType()
export default class Game extends BaseEntity {
  @PrimaryGeneratedColumn("uuid")
  @Field(() => ID)
  id: string;

  @Column()
  @Field()
  name: string;

  @ManyToOne(() => User)
  @Field()
  master: User;

  @OneToMany(() => GameUser, (gameUser) => gameUser.game)
  gameUsers: GameUser[];

  @Field(() => [User])
  users(): User[] {
    return this.gameUsers.map((gu) => gu.user);
  }

  @Column({ nullable: true })
  @Field({ nullable: true })
  password: string;

  @CreateDateColumn({ type: "timestamp" })
  @Field()
  createdAt: Date;

  @UpdateDateColumn({ type: "timestamp" })
  @Field()
  updatedAt: Date;
}
