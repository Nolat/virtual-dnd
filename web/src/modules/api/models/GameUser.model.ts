import { Field, ID, ObjectType } from "type-graphql";
import { BaseEntity, Column, Entity, ManyToOne, PrimaryGeneratedColumn } from "typeorm";

import { Game, User } from "modules/api/models";

@Entity()
@ObjectType()
export default class GameUser extends BaseEntity {
  @PrimaryGeneratedColumn("uuid")
  @Field(() => ID)
  id: string;

  @ManyToOne(() => User, (user) => user.games)
  @Field(() => User)
  user: User;

  @ManyToOne(() => Game, (game) => game.users)
  @Field()
  game: Game;

  @Column({ type: "timestamp", nullable: true })
  @Field({ nullable: true })
  lastSeenAt: Date;
}
