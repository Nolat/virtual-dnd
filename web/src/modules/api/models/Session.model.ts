import { randomBytes } from "crypto";

import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  OneToOne,
  PrimaryGeneratedColumn,
  RelationId,
  UpdateDateColumn
} from "typeorm";

import { User } from "modules/api/models";

@Entity()
export default class Session extends BaseEntity {
  constructor() {
    super();
    this.sessionToken = randomBytes(32).toString("hex");
    this.accessToken = randomBytes(32).toString("hex");
  }

  @PrimaryGeneratedColumn("uuid")
  id: string;

  @OneToOne(() => User)
  @JoinColumn()
  user: User;

  @RelationId((session: Session) => session.user)
  userId: string;

  @Column({ type: "timestamp" })
  expiresOn: Date;

  @Column({ unique: true })
  sessionToken: string;

  @Column({ unique: true })
  accessToken: string;

  @CreateDateColumn({ type: "timestamp" })
  createdAt: Date;

  @UpdateDateColumn({ type: "timestamp" })
  updatedAt: Date;
}
