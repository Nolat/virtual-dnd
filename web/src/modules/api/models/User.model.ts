import { Field, ID, ObjectType } from "type-graphql";
import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn
} from "typeorm";

import { Game, GameUser } from "modules/api/models";

@Entity()
@ObjectType()
export default class User extends BaseEntity {
  @PrimaryGeneratedColumn("uuid")
  @Field(() => ID)
  id: string;

  @Column()
  @Field()
  name: string;

  @Column({ unique: true, nullable: true })
  @Field()
  email: string;

  @Column()
  @Field()
  image: string;

  @OneToMany(() => GameUser, (gameUser) => gameUser.user)
  gameUsers: GameUser[];

  @Field(() => [Game], { nullable: true })
  games(): Game[] {
    return this.gameUsers.map((gu) => gu.game);
  }

  @CreateDateColumn({ type: "timestamp" })
  @Field()
  createdAt: Date;

  @UpdateDateColumn({ type: "timestamp" })
  @Field()
  updatedAt: Date;
}
