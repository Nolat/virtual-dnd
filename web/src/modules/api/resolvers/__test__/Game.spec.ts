import { gql } from "apollo-server-micro";
import { createTestClient } from "apollo-server-testing";
import { getConnection } from "typeorm";

import { Game, GameUser, User } from "modules/api/models";
import { bootstrapTestServer } from "test/setup/bootstrapServer";

describe("Game resolver", () => {
  const GAME_DATA = {
    name: "Test Game",
    password: "Test"
  };

  const USER_DATA = {
    email: "test@test.com",
    name: "test-user",
    image: "img:test"
  };

  const USER2_DATA = {
    email: "test2@test.com",
    name: "test2-user",
    image: "img:test2"
  };

  it("should get a specific game from id", async () => {
    const connection = getConnection();
    const gameRepository = connection.getRepository(Game);
    const gameUserRepository = connection.getRepository(GameUser);
    const userRepository = connection.getRepository(User);

    const user = userRepository.create(USER_DATA);
    await user.save();

    const game = gameRepository.create({ ...GAME_DATA, master: user });
    await game.save();

    const gameUser = gameUserRepository.create({ user, game });
    await gameUser.save();

    const server = await bootstrapTestServer({});

    const { query } = createTestClient(server);

    const res = await query({
      query: gql`
        query Game($id: String!) {
          Game(id: $id) {
            name
            master {
              name
            }
            users {
              name
            }
          }
        }
      `,
      variables: { id: game.id }
    });

    expect(res.data.Game).toMatchSnapshot();
  });

  it("should create a new game", async () => {
    const connection = getConnection();
    const userRepository = connection.getRepository(User);

    const user = await userRepository.create(USER_DATA);
    await user.save();

    const server = await bootstrapTestServer({
      context: async () => {
        return { user };
      }
    });

    const { mutate } = createTestClient(server);

    const res = await mutate({
      mutation: gql`
        mutation CreateGame($name: String!, $password: String) {
          CreateGame(name: $name, password: $password) {
            name
            master {
              name
            }
            users {
              name
            }
          }
        }
      `,
      variables: GAME_DATA
    });

    expect(res.data.CreateGame).toMatchSnapshot();
  });

  it("should prevent creating a game when user is not in context", async () => {
    const server = await bootstrapTestServer({});

    const { mutate } = createTestClient(server);

    const res = await mutate({
      mutation: gql`
        mutation CreateGame($name: String!, $password: String) {
          CreateGame(name: $name, password: $password) {
            name
          }
        }
      `,
      variables: GAME_DATA
    });

    expect(res.errors).toMatchSnapshot();
  });

  it("should get joining game information", async () => {
    const connection = getConnection();
    const gameRepository = connection.getRepository(Game);
    const gameUserRepository = connection.getRepository(GameUser);
    const userRepository = connection.getRepository(User);

    const user = userRepository.create(USER_DATA);
    await user.save();

    const master = userRepository.create(USER2_DATA);
    await master.save();

    const game = gameRepository.create({ ...GAME_DATA, master });
    await game.save();

    const server = await bootstrapTestServer({
      context: async () => {
        return { user };
      }
    });

    const { query } = createTestClient(server);

    const JoinGameInfoQuery = gql`
      query JoinGameInfo($id: String!) {
        JoinGameInfo(id: $id) {
          hasJoined
          hasPassword
        }
      }
    `;

    const res1 = await query({
      query: JoinGameInfoQuery,
      variables: { id: game.id }
    });

    expect(res1.data.JoinGameInfo).toMatchSnapshot();

    const gameUser = gameUserRepository.create({ user, game });
    await gameUser.save();

    const res2 = await query({
      query: JoinGameInfoQuery,
      variables: { id: game.id }
    });

    expect(res2.data.JoinGameInfo).toMatchSnapshot();

    game.password = null;
    await game.save();

    const res3 = await query({
      query: JoinGameInfoQuery,
      variables: { id: game.id }
    });

    expect(res3.data.JoinGameInfo).toMatchSnapshot();
  });

  it("should add user to game (without password)", async () => {
    const connection = getConnection();
    const userRepository = connection.getRepository(User);
    const gameRepository = connection.getRepository(Game);
    const gameUserRepository = connection.getRepository(GameUser);

    const master = userRepository.create(USER2_DATA);
    await master.save();

    const game = gameRepository.create({ ...GAME_DATA, password: null, master });
    await game.save();

    const gameUser = gameUserRepository.create({ user: master, game });
    await gameUser.save();

    const user = await userRepository.create(USER_DATA);
    await user.save();

    const server = await bootstrapTestServer({
      context: async () => {
        return { user };
      }
    });

    const { mutate } = createTestClient(server);

    const JoinGameMutation = gql`
      mutation JoinGame($id: String!) {
        JoinGame(id: $id) {
          name
          master {
            name
          }
          users {
            name
          }
        }
      }
    `;

    const res1 = await mutate({
      mutation: JoinGameMutation,
      variables: { id: game.id.replace(/.$/, "1") }
    });

    expect(res1.errors).toMatchSnapshot();

    const res2 = await mutate({
      mutation: JoinGameMutation,
      variables: { id: game.id }
    });

    expect(res2.data).toMatchSnapshot();
  });

  it("should add user to game (with password)", async () => {
    const connection = getConnection();
    const userRepository = connection.getRepository(User);
    const gameRepository = connection.getRepository(Game);
    const gameUserRepository = connection.getRepository(GameUser);

    const master = userRepository.create(USER2_DATA);
    await master.save();

    const game = gameRepository.create({ ...GAME_DATA, master });
    await game.save();

    const gameUser = gameUserRepository.create({ user: master, game });
    await gameUser.save();

    const user = await userRepository.create(USER_DATA);
    await user.save();

    const server = await bootstrapTestServer({
      context: async () => {
        return { user };
      }
    });

    const { mutate } = createTestClient(server);

    const JoinGameMutation = gql`
      mutation JoinGame($id: String!, $password: String) {
        JoinGame(id: $id, password: $password) {
          name
          master {
            name
          }
          users {
            name
          }
        }
      }
    `;

    const res1 = await mutate({
      mutation: JoinGameMutation,
      variables: { id: game.id }
    });

    expect(res1.errors).toMatchSnapshot();

    const res2 = await mutate({
      mutation: JoinGameMutation,
      variables: { id: game.id, password: GAME_DATA.password }
    });

    expect(res2.data).toMatchSnapshot();
  });

  it("should remove user from game (as player)", async () => {
    const connection = getConnection();
    const userRepository = connection.getRepository(User);
    const gameRepository = connection.getRepository(Game);
    const gameUserRepository = connection.getRepository(GameUser);

    const user1 = userRepository.create(USER_DATA);
    await user1.save();

    const game = gameRepository.create({ ...GAME_DATA, master: user1 });
    await game.save();

    const gameUser1 = gameUserRepository.create({ user: user1, game });
    await gameUser1.save();

    const user2 = await userRepository.create(USER2_DATA);
    await user2.save();

    const gameUser2 = gameUserRepository.create({ user: user2, game });
    await gameUser2.save();

    const server = await bootstrapTestServer({
      context: async () => {
        return { user: user2 };
      }
    });

    const { mutate } = createTestClient(server);

    const res = await mutate({
      mutation: gql`
        mutation LeaveGame($id: String!) {
          LeaveGame(id: $id)
        }
      `,
      variables: { id: game.id }
    });

    expect(res.data.LeaveGame).toStrictEqual(true);
  });

  it("should prevent user (gm) from leaving game", async () => {
    const connection = getConnection();
    const userRepository = connection.getRepository(User);
    const gameRepository = connection.getRepository(Game);
    const gameUserRepository = connection.getRepository(GameUser);

    const user = userRepository.create(USER_DATA);
    await user.save();

    const game = gameRepository.create({ ...GAME_DATA, master: user });
    await game.save();

    const gameUser = gameUserRepository.create({ user, game });
    await gameUser.save();

    const server = await bootstrapTestServer({
      context: async () => {
        return { user };
      }
    });

    const { mutate } = createTestClient(server);

    const res = await mutate({
      mutation: gql`
        mutation LeaveGame($id: String!) {
          LeaveGame(id: $id)
        }
      `,
      variables: { id: game.id }
    });

    expect(res.data.LeaveGame).toStrictEqual(false);
  });
});
