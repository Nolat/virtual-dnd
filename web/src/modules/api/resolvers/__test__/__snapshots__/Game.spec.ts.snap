// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game resolver should add user to game (with password) 1`] = `
Array [
  [GraphQLError: Wrong password],
]
`;

exports[`Game resolver should add user to game (with password) 2`] = `
Object {
  "JoinGame": Object {
    "master": Object {
      "name": "test2-user",
    },
    "name": "Test Game",
    "users": Array [
      Object {
        "name": "test2-user",
      },
      Object {
        "name": "test-user",
      },
    ],
  },
}
`;

exports[`Game resolver should add user to game (without password) 1`] = `
Array [
  [GraphQLError: Game not found],
]
`;

exports[`Game resolver should add user to game (without password) 2`] = `
Object {
  "JoinGame": Object {
    "master": Object {
      "name": "test2-user",
    },
    "name": "Test Game",
    "users": Array [
      Object {
        "name": "test2-user",
      },
      Object {
        "name": "test-user",
      },
    ],
  },
}
`;

exports[`Game resolver should create a new game 1`] = `
Object {
  "master": Object {
    "name": "test-user",
  },
  "name": "Test Game",
  "users": Array [
    Object {
      "name": "test-user",
    },
  ],
}
`;

exports[`Game resolver should get a specific game from id 1`] = `
Object {
  "master": Object {
    "name": "test-user",
  },
  "name": "Test Game",
  "users": Array [
    Object {
      "name": "test-user",
    },
  ],
}
`;

exports[`Game resolver should get joining game information 1`] = `
Object {
  "hasJoined": false,
  "hasPassword": true,
}
`;

exports[`Game resolver should get joining game information 2`] = `
Object {
  "hasJoined": true,
  "hasPassword": true,
}
`;

exports[`Game resolver should get joining game information 3`] = `
Object {
  "hasJoined": true,
  "hasPassword": false,
}
`;

exports[`Game resolver should prevent creating a game when user is not in context 1`] = `
Array [
  [GraphQLError: Access denied! You need to be authorized to perform this action!],
]
`;
